### ISSUE TITLE
HEAD does not build with -mod=vendor

### ISSUE THREAD

**Issue by querry43** [hashicorp/terraform#20208]
### Terraform Version

```
commit d1daa7e5187d0120620f260ec1aed9631b77f6f8 (HEAD -> master, origin/master, origin/HEAD)
Author: Nick Fagerlund <nick.fagerlund@gmail.com>
Date:   Fri Jan 18 18:21:30 2019 -0800

    website: Make 0.11 and 0.12 subsections of 'Configuration Language'
```

```
matt@v1:~/work/terraform$ go version
go version go1.11.5 linux/amd64
```

### Terraform Configuration Files

```hcl
<none>
```

### Debug Output

```
matt@v1:~/work/terraform$ make
==> Checking that code complies with gofmt requirements...
GO111MODULE=off go get -u golang.org/x/tools/cmd/stringer
GO111MODULE=off go get -u golang.org/x/tools/cmd/cover
GO111MODULE=off go get -u github.com/golang/mock/mockgen
# We turn off modules for "go generate" because our downstream generate
# commands are not all ready to deal with Go modules yet.
# See https://github.com/golang/go/issues/24661
# This also avoids downloading all of the deps that are in the vendor dir anyway.
GO111MODULE=off go generate ./...
2019/02/02 20:33:04 Generated command/internal_plugin_list.go
GO111MODULE=off go fmt command/internal_plugin_list.go > /dev/null
go list -mod=vendor ./... | xargs -t -n4 go test  -mod=vendor -timeout=2m -parallel=4
build flag -mod=vendor only valid when using modules
go test -mod=vendor -timeout=2m -parallel=4
build flag -mod=vendor only valid when using modules
Makefile:35: recipe for target 'test' failed
make: *** [test] Error 123
```

### Expected Behavior

go test and go install run successfully when called by `make` and `make dev`

### Actual Behavior

go test and go install complain about -mod=vendor and exit with code 1

### Steps to Reproduce

1. `make`

### Additional Context

This patch lets me complete the build.

```
diff --git Makefile Makefile
index 03be211e5..45515a4ae 100644
--- Makefile
+++ Makefile
@@ -17,7 +17,7 @@ bin: fmtcheck generate
 # dev creates binaries for testing Terraform locally. These are put
 # into ./bin/ as well as $GOPATH/bin
 dev: fmtcheck generate
-       go install -mod=vendor .
+       go install .

 quickdev: generate
        go install -mod=vendor .
@@ -32,7 +32,7 @@ plugin-dev: generate
 # we run this one package at a time here because running the entire suite in
 # one command creates memory usage issues when running in Travis-CI.
 test: fmtcheck generate
-       go list -mod=vendor $(TEST) | xargs -t -n4 go test $(TESTARGS) -mod=vendor -timeout=2m -parallel=4
+       go list -mod=vendor $(TEST) | xargs -t -n4 go test $(TESTARGS) -timeout=2m -parallel=4

 # testacc runs acceptance tests
 testacc: fmtcheck generate
```

**Comment by apparentlymart** [comment]
Hi @querry43,

Sorry for the unhelpful failure here. Terraform expects to be build with Go 1.11.5 in modules mode. To achieve that you can either check it out somewhere outside your `GOPATH` or set the environment variable ` GO111MODULES=on` to force it.

We'll use this issue to represent seeing if we can arrange for the makefile to force modules mode itself, avoid that additional step. We also have some docs updating to do; they are lagging a bit because we usually do most updating in conjunction with preparing a release, but we haven't done one for a while.

**Comment by querry43** [comment]
Thanks @apparentlymart , makes sense.

**Comment by bentterp** [comment]
Note to myself mainly: when Martin writes "set the environment variable ` GO111MODULES=on` to force it", it means to run this command: ```GO111MODULE=on make``` or ```GO111MODULE=on make dev```
Took me a while to figure out but admittedly, I'm a bit slow.... 😊

**Comment by bentterp** [comment]
Sorry, but I can't get it to work:
```
:~/.local/go/src/github.com/hashicorp/terraform$ echo $GO111MODULES $GO111MODULE
on on
:~/.local/go/src/github.com/hashicorp/terraform$ GO111MODULE=on GO111MODULES=on make dev
==> Checking that code complies with gofmt requirements...
GO111MODULE=off go get -u golang.org/x/tools/cmd/stringer
GO111MODULE=off go get -u golang.org/x/tools/cmd/cover
GO111MODULE=off go get -u github.com/golang/mock/mockgen
GOFLAGS=-mod=vendor go generate ./...
2019/02/21 14:42:34 Generated command/internal_plugin_list.go
# go fmt doesn't support -mod=vendor but it still wants to populate the
# module cache with everything in go.mod even though formatting requires
# no dependencies, and so we're disabling modules mode for this right
# now until the "go fmt" behavior is rationalized to either support the
# -mod= argument or _not_ try to install things.
GO111MODULE=off go fmt command/internal_plugin_list.go > /dev/null
go install -mod=vendor .
:~/.local/go/src/github.com/hashicorp/terraform$ git log -n 1
commit 4ab701620e4e28950917b8407babb67c4cabf3fc (HEAD -> master, origin/master, origin/HEAD)
Merge: 085ac6d8c 876d548bc
Author: Radek Simko <radek.simko@gmail.com>
Date:   Thu Feb 21 07:29:12 2019 +0000

    Merge pull request #20326 from hashicorp/vendor-grpc-go-bump

    vendor: Bump google.golang.org/grpc & cloud.google.com/go & oauth2
:~/.local/go/src/github.com/hashicorp/terraform$
```

But ```make bin``` so that's something

**Comment by apparentlymart** [comment]
Hi @bentterp!

I don't see any failure in the output you shared there... it looks like it actually worked. The comments in the makefile generate some noise in the output, but `go install -mod=vendor .` seems to have succeeded without errors, so you should find the built Terraform binary in your `GOBIN` directory.

**Comment by apparentlymart** [comment]
That error message indicates that the `azurerm` plugin doesn't have support for Terraform v0.12.0. There's a better error message for that situation now in the case where plugins are being installed from `terraform init`, but I think in that case `terraform init` found a local `terraform-provider-azurerm-...` file and tried to use it.

As a stopgap until we have a criticial mass of providers released with v0.12.0 support, we've uploaded [some dev snapshots](http://terraform-0.12.0-dev-snapshots.s3-website-us-west-2.amazonaws.com/) which are built from latest master while forcing the providers to use the latest Terraform SDK with v0.12.0 support (as I write this, the newest builds are now a week old, but should still work). If you use those alongside a build from Terraform latest master (or, alternatively, use the dev snapshot build of Terraform too) then you should find a working configuration.

There will be official releases of providers with v0.12.0 support landing shortly, once the individual provider teams complete any required updates and testing. At that point `terraform init` should start working again to install those providers.

**Comment by apparentlymart** [comment]
Hi again, all!

From reviewing the above it seems like the original issue is now resolved by noting that Terraform requires Go modules. The next version of Go (this month) will enable modules mode by default, which will hopefully resolve any remaining friction here.

I'm going to close this out now. If you encounter any other build-related issues, please feel free to open a new issue. Thanks for reporting this, @querry43.
